src/main/kotlin/org/elm/lang/core/psi/ElmFile.kt

```diff
@@ -124,3 +136,20 @@ class ElmFile(viewProvider: FileViewProvider) : PsiFileBase(viewProvider, ElmLan
 
 val VirtualFile.isElmFile
     get() = fileType == ElmFileType
+
+
+/**
+ * Describes the location from which a reference is resolved.
+ *
+ * Reference resolution depends on context. For instance, we need
+ * to know the containing Elm project in order to determine which
+ * `source-directories` are valid roots.
+ *
+ * Starting in Elm 0.19, the Elm project's `test-dependencies`
+ * are only "import-able" from within "$ProjectRoot/tests" directory.
+ */
+interface ClientLocation {
```

Comment:
Not sure about this name. Maybe it would be better as `RefContext`?

I also wonder if it's a bit too clever making `ElmFile` implement this interface. Maybe it'd be better to make things more explicit by giving `ClientLocation` a factory function that takes an `ElmFile`.